openmicroscopy/openmicroscopy
# Api child count

# What this PR does

 - Uses "data" wrapper on all json returned from GET, POST & PUT for single object
 - Updates version to ```0```. E.g. ```/api/v0/m/screens/```
 - Uses separate queries for loading childCount
 - Adds childCount support for single objects, e.g.  ```/api/v0/m/projects/:id/```

# Testing this PR

1. Integration tests are updated to handle "data" wrapper and to test childCount for Projects, Datasets and Screens (both single P/D/S and lists).
 2. Manually testing:
    - browse to E.g. ```/api/v0/m/projects/:id/``` and check that project json is within a ```"data": {}``` wrapper.
    - Check that adding ```?childCount=true``` adds an ```omero:childCount``` to the Project json.
    - ```?childCount=true``` should also work for /projects/, /datasets/, /screens/ and for single /datasets/:id and /screens/:id/


# Related reading

This PR removes child_count support from ```conn.buildQuery()``` and instead this is provided by the web api itself in ```api_query.get_child_counts()```. This is partly due to the small performance improvement (see trello card) but also because it makes the  ```conn.buildQuery()``` more robust. Previously if child_count was True, this returned a query that required ```queryService.projection()``` instead of simply ```queryService.findAllByQuery()```.

1. https://trello.com/c/mN819Y4B/263-childcount-performance
